File: Python/Types of Matrix/Skew-Hermithian matrix.py
# Python ≥3.10 is required to run this program
import re
def isNumeric(value):
    try:
        int(value)
        return True
    except:
        return False


def isOperator(operator):
    if operator == '+' or operator == '-':
        return True
    else:
        return False


def conjugate(expression, call):
    chs = re.split(r'(?=[-+*/])', expression.strip())
    sn = 0  # sum of numerics
    expression = ""
    for i in range(0, len(chs)):
        val = chs[i]
        if len(val) > 0:
            operator = None
            if isOperator(val[0]):
                operator = val[0]
                val = val[1:]
            elif (i == 0):
                operator = '+'

            if len(re.sub(r'^(\d*?[a-zA-Z$])', "", val)) > 0 and isNumeric(val) == False:
                return "error"
            else:
                if isNumeric(val) and call == True: # default expression only
                    if int(val) != 0:
                        match operator:
                            case '+': sn += int(val)
                            case '-': sn -= int(val)
                else: # default expression and nagetive multiplication
                    match operator:
                        case '+': expression += '-' + val
                        case '-': expression += '+' + val

    if sn == 0 and len(expression) == 0:
        expression = str(sn)
    elif sn != 0:
        expression = str(sn) + expression

    if isOperator(expression[0]) and (expression[0] == '+' or expression == "-0"):
        expression = expression[1:]

    return expression.strip()


A = [
    ["-i", "2+i"],
    ["-2+i", "0"]
]

B = [[None]*len(A) for row in range(len(A[0]))]  # A^H or A^†
C = [[None]*len(A[0]) for row in range(len(A))]  # -A

print("Main Matrix (A): ")
for row in range(0, len(A)):
    for col in range(0, len(A[row])):
        print("\t", A[row][col], "\t", end='')
        B[col][row] = conjugate(A[row][col], True); # true => will return default processed expression 
        C[row][col] = conjugate(A[row][col], False); # false => will return nagetive multiplication expression, or (-A). [Note: Multiplication used on the same method to reduce the codes ]
    print()


# Check if any element/expression has error
if len(A) == len(A[0]):
    error = False; hermithian = True
    for row in range(0, len(B)):
        for col in range(0, len(B[row])):
            if B[row][col] == "error" and error == False:
                error = True
            if A[row][col] != B[row][col] and hermithian == True:
                hermithian = False


    if error:
        print("\nError! Invalid expression used.")
    else:
        skew_validation = True
        print("\n╔══════════════════════════════════════════╗");
        print("  A^†: ")
        for row in range(0, len(B)):
            for col in range(0, len(B[row])):
                print("\t", B[row][col], "\t", end='')
                if B[row][col] != C[row][col] and skew_validation == True:
                    skew_validation = False
            print()
        
        print("╠══════════════════════════════════════════╣")

        print("  -A: ");
        for row in range(0, len(C)):
            for col in range(0, len(C[row])):
                print("\t", C[row][col], "\t", end='')
            print()

        print("╚══════════════════════════════════════════╝")
        
        if skew_validation:
            print("↳ (A^† = -A) Therefore, it's a Skew-Hermithian matrix")
        else:
            print("↳ (A^† ≠ -A) Therefore, it's not a Skew-Hermithian matrix")

else:
    print("\nError! The main matrix must be an square to be a Skew-hermethian matrix.")

File: C++/Types of Matrix/Skew-Hermithian matrix.cpp
#include <iostream>
#include <vector>
#include <regex>
using namespace std;

vector <string> split(string content, regex match){
    vector <string> elements;
    regex_token_iterator<string::iterator>
        iterator(content.begin(), content.end(), match, -1);
    regex_token_iterator<string::iterator> end;
    while (iterator != end)
        elements.push_back(*iterator++);
    return elements;
}
bool isNumeric(string value){
    return value.find_first_not_of("0123456789") == string::npos;
}
bool isOperator(char Operator){
    if(Operator == '+' || Operator == '-'){
        return true;
    }else{
        return false;
    }
}
string conjugate(string expression, bool call){
    vector<string> chs = split(expression, regex("(?=[-+*/])"));
    int sn = 0; // sum of numerics
    expression = "";
    for (int i = 0; i < chs.size(); i++) {
        string val = chs[i];
        if(val.length() > 0){
            char Operator = 0;
            if(isOperator(val.at(0))){
                Operator = val.at(0);
                val = val.substr(1);
            } else if (i == 0) {
                Operator = '+';
            }

            if(regex_replace(val, regex("^(\\d*?[a-zA-Z$])"), "").length() > 0 && !isNumeric(val)){
                return "error";
            }else{
                if(isNumeric(val) && call == true){ // default expression only
                    if(stoi(val) != 0){
                        switch (Operator) {
                            case '+': sn += stoi(val); break;
                            case '-': sn -= stoi(val); break;
                        }
                    }
                } else {  // default expression and nagetive multiplication
                    switch (Operator) {
                        case '+': expression += '-' + val; break;
                        case '-': expression += '+' + val; break;
                    }
                }
            }
        }
    }
    if(sn == 0 && expression.length() == 0){
        expression = to_string(sn);
    }else if(sn != 0){
        expression = to_string(sn).append(expression);
    }
    if (isOperator(expression.at(0)) && (expression.at(0) == '+' || expression == "-0")) {
        expression = expression.substr(1);
    }
    return regex_replace(expression, regex("^\\s+|\\s+$"), "");
}

int main(){
    vector<vector<string>> A = {
        { "-i", "2+i" },
        { "-2+i", "0" }
    };

    vector <vector<string>> B (A[0].size(), vector <string> (A.size())); // A^H or A^†
    vector <vector<string>> C (A.size(), vector <string> (A[0].size())); // -A

    cout << "Main Matrix (A): " << endl;
    for (int row = 0; row < A.size(); row++) {
        for (int col = 0; col < A[row].size(); col++) {
            cout << "\t" << A[row][col] << "\t";
            B[col][row] = conjugate(A[row][col], true); // true => will return default processed expression 
            C[row][col] = conjugate(A[row][col], false); // false => will return nagetive multiplication expression, or (-A). [Note: Multiplication used on the same method to reduce the codes]
        }
        cout << endl;
    }

    // Check if any element/expression has error
    if(A.size() == A[0].size()){
        bool error = false;  bool hermithian = true;
        for (int row = 0; row < B.size(); row++) {
            for (int col = 0; col < B[row].size(); col++) {
                if(B[row][col] == "error" && error == false) error = true;
                if( A[row][col] != B[row][col] && hermithian == true ){
                    hermithian = false;
                }
            }
        }
    
        if(error){
            cout << "\nError! Invalid expression used." << endl;
        } else {
            bool skew_validation = true;
            cout << "\n╔══════════════════════════════════════════╗" << endl;
            cout << "  A^†: " << endl;
            for (int row = 0; row < B.size(); row++) {
                for (int col = 0; col < B[row].size(); col++) {
                    cout << "\t" + B[row][col] + "\t";
                    if (B[row][col] != C[row][col] && skew_validation == true) {
                        skew_validation = false;
                    }
                }
                cout << endl;
            }
            
            cout << "╠══════════════════════════════════════════╣" << endl;

            cout << "  -A: " << endl;
            for (int row = 0; row < C.size(); row++) {
                for (int col = 0; col < C[row].size(); col++) {
                    cout << "\t" + C[row][col] + "\t";
                }
                cout << endl;
            }
            cout << "╚══════════════════════════════════════════╝" << endl;
            
            if (skew_validation) {
                cout << "↳ (A^† = -A) Therefore, it's a Skew-Hermithian matrix" << endl;
            } else {
                cout << "↳ (A^† ≠ -A) Therefore, it's not a Skew-Hermithian matrix" << endl;
            }
        }

    } else {
        cout << "\nError! The main matrix must be an square to be a Skew-hermethian matrix." << endl;
    }
}